app/controllers/concerns/find_common.rb

```diff
@@ -10,23 +10,39 @@ def find_resource
   end
 
   def resource_name
-    controller_name.singularize
+    resource_name_for(controller_name)
   end
 
   def resource_class
-    @resource_class ||= resource_name.classify.constantize
+    @resource_class ||= resource_class_for(resource_name)
   end
 
   def resource_scope(options = {})
-    @resource_scope ||= begin
-      options[:controller] ||= controller_name
-      options[:permission] ||= "#{action_permission}_#{options[:controller]}"
-      scope = resource_class.scoped
-      if resource_class.respond_to?(:authorized)
-        scope.authorized(options[:permission], resource_class)
-      else
-        scope
-      end
+    @resource_scope ||= scope_for(resource_class, options)
+  end
+
+  def scope_for(resource, options = {})
+    controller = options.delete(:controller){ controller_name }
+    #don't call the #action_permission method here, we are not sure if the resource is authorized at this point
+    permission = options.delete(:permission)
+
+    if resource.respond_to?(:authorized)
+      permission ||= "#{action_permission}_#{controller}"
+      resource = resource.authorized(permission, resource)
+    end
+
+    resource.where(options)
+  end
+
+  def resource_name_for(resource)
+    resource.singularize
+  end
+
+  def resource_class_for(resource)
+    begin
+      return resource.classify.constantize
+    rescue NameError
```

Comment:
That was my point, this makes parent_class equals to nil there which is passed to `scope_for(parent_class,` which can't handle nil as parent_class.

What's more important you refactored `resource_class` to use `resource_class_for` as well that is used in `resource_scope` like this `association.authorized(options[:permission], resource_class)` so you'd allow to pass nil as a class to authorized which propagates to https://github.com/theforeman/foreman/blob/develop/app/models/concerns/authorizable.rb#L21. That would cause to check permissions on wrong resource.

Maybe it would be better to find class name by searching through reflections with such foreign key rather then trying to constantize all parameters.